Shoddy search

jaredonline opened the pull request:
This is literally the first time I've ever written Hack, so I'm sure I did lots wrong.

Please be merciless in your feedback =D

Signed-off-by: Jared McFarland <jared.online@gmail.com>

fredemmott commented on LocalConfig.php.example:
```diff
@@ -1,8 +0,0 @@
-abstract final class LocalConfig {
-  const string BUILD_DIR = __DIR__.'/build';
-  const string HHVM_TREE = '/home/fred/hhvm';
-}
```
?

JoelMarcey replied:
Yep. Keep this file.

fredemmott commented on md-render/Gemfile.lock:
```diff
@@ -35,6 +35,3 @@ DEPENDENCIES
   pygments.rb
-
-BUNDLED WITH
-   1.10.6
```
?

jazzdan commented on LocalConfig.php.example:
```diff
@@ -1,8 +0,0 @@
-<?hh // strict
```
We prolly don't want to delete this. :)

fredemmott commented on src/site/ParamGetter.php:
```diff
@@ -0,0 +1,42 @@
+<?hh
+
+// a wrapper around the 'unstrict' $_GET and $_POST vars
+final class ParamGetter {
```
We have this via the PSR7 request interface, and get*Param functions in WebPageController. See changes to WebController and getRequiredStringParam() in WebPageContentController in https://github.com/hhvm/user-documentation/commit/955072a71b5405d2db52de8e8e145a4014c9f398 for an example

fredemmott replied:
You'll need to extend WebPageController in a similar manner to support $_POST

fredemmott commented on src/site/SearchController.php:
```diff
@@ -0,0 +1,71 @@
+  public function __construct(ImmMap<string, string> $params) {
+    parent::__construct($params);
+    $this->term = "";
+    $this->term_set = false;
```
- this constructor is no longer valid - hh_client should complain
- not actually needed - could use static initializers instead
- why $term_set = false instead of ?string $term = null

fredemmott commented on src/site/SearchController.php:
```diff
@@ -0,0 +1,71 @@
+        $list->appendChild($item);
+    }
+    if (count($result_set) === 0) {
+        $item = <li><em>No results for this category</em></li>;
```
Nit: bit weird to still use a `<ul><li>` for this - could some other markup be better?

fredemmott commented on src/site/SearchController.php:
```diff
@@ -0,0 +1,71 @@
+        You searched for "{$this->getSearchTerm()}"
+        <h2>Guides</h2>
+        <h3>Hack</h3>
+            {$this->getListFromResultSet($results->getHackGuides())}
```
- 2 space indentation, not 4
 - not actually a child, shouldn't be indented

fredemmott commented on src/site/SearchController.php:
```diff
@@ -0,0 +1,71 @@
+
+  private function getSearchTerm(): string {
+    if ($this->term_set) {
+        return $this->term;
```
2space

JoelMarcey commented on src/GuidesIndex.php:
```diff
@@ -25,6 +25,24 @@ private static function getSummaries(
     return require(BuildPaths::GUIDES_SUMMARY);
   }
 
+  public static function search(string $term, SearchResultSet &$results): void {
```
Couldn't you just return a `SearchResultSet` instead of using a reference?

fredemmott replied:
No references in hack.

Alternatively, return a vector of results, or if you have different tyeps of results, return your own searchresultset, and do something like:

$combined_results->addResults(Foo::search($term));

JoelMarcey replied:
Or wait, won't this be returned by reference anyway?

SiebelsTim replied:
Here too. 
Also references are banned in strict mode. (Which this file sadly isn't :()

fredemmott replied:
yeah, objects are byref (in PHP too), so the `&` is unneeded - still, the other suggested approaches here would be cleaner than depending on that.

SiebelsTim commented on src/APIIndex.php:
```diff
@@ -24,6 +24,17 @@ public static function getReferenceForType(
     return $index[$type];
   }
 
+  public static function search(string $term, SearchResultSet &$results): void {
```
Objects are passed by reference by default.

fredemmott commented on src/site/SearchResultSet.php:
```diff
@@ -0,0 +1,60 @@
+        $this->interfaces = Map {};
+        $this->hhvm_guides = Map {};
+        $this->hack_guides = Map {};
+    }
```
prefer static initializers to constructor

fredemmott commented on src/site/SearchResultSet.php:
```diff
@@ -0,0 +1,60 @@
+
+    public function addAPIResult(string $type, string $name): void {
+        if ($type === "class") {
+            $this->classes[$name] = sprintf("/hack/reference/class/%s/", $name);
```
doesn't need fixing for this, but FYI, I mean to add a URI builder - so something like `GuidePageController::getURIBuilder()->setStringParam('class', $name)->GetURI()` to avoid hardcoded paths

fredemmott commented on src/site/SearchResultSet.php:
```diff
@@ -0,0 +1,60 @@
+        }
+        if ($type === "interface") {
+            $this->interfaces[$name] = sprintf("/hack/reference/interface/%s/", $name);
+        }
```
has this been tested with namespaced stuff? `HH\Asio\null()` in particular can be fun

jaredonline replied:
It works fine because the pathing uses `.` instead of `\`, for example: `/hack/reference/function/HH.Asio.null/`

JoelMarcey commented:
Thanks for the pull request @jaredonline !! 

:+1: 

I hope you feel most of the comments are constructive (I hope).

SiebelsTim commented on src/site/SearchController.php:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  protected async function getTitle(): Awaitable<string> {
+    return 'Search for shit!';
```
Ahem.

fredemmott replied:
:poop:

JoelMarcey replied:
LOL

jaredonline commented:
Hey folks! Sorry it took me so long \= I was having a really hard time getting it to run on my local laptop (outside the custom environment @jazzdan setup for us at Etsy).

I'm pretty sure I've addressed all the feedback at this point.

jazzdan commented:
![https://media.riffsy.com/images/eda65eadf312e50e52406da4f51a2906/raw](https://media.riffsy.com/images/eda65eadf312e50e52406da4f51a2906/raw)

fredemmott commented on .gitignore:
```diff
@@ -10,6 +10,7 @@ LocalConfig.php
 *.css.map
 /build/
 /public/www.pid
+/.sass-cache/
```
?

jaredonline replied:
I think I had an old `.sass-cache` dir laying around and I added this so I didn't accidentally push it. I wasn't sure why the sass stuff all disappeared on me, but it looks like it was removed intentionally. I'll remove my local dir and this line.

fredemmott commented on src/APIIndex.php:
```diff
@@ -27,6 +27,23 @@ public static function getReferenceForType(
+    // This whole method is UNSAFE
+    $index = Shapes::toArray(self::getIndex());
+    foreach ($index as $key => $value) {
+        if (is_array($value)) {
```
2space indents

fredemmott commented on src/APIIndex.php:
```diff
@@ -27,6 +27,23 @@ public static function getReferenceForType(
+            foreach ($value as $name => $entry) {
+                if (is_string($name) && is_string($key)) {
+                    if (strtolower($name) === strtolower($term) || strpos(strtolower($name), strtolower($term))) {
+                        $results->addAPIResult($key, $name);
```
you might want to also special-case dealing with "HH\" here

jaredonline replied:
What would the special case be? I haven't run into any issues with it so far.

fredemmott commented on src/site/SearchController.php:
```diff
@@ -0,0 +1,68 @@
+  private ?string $term = null;
+
+  public async function getTitle(): Awaitable<string> {
+    return 'Search for shit!';
```
:poop:

fredemmott commented on src/site/SearchResultSet.php:
```diff
@@ -0,0 +1,52 @@
+    private Map<string, string> $hhvm_guides = Map {};
+    private Map<string, string> $hack_guides = Map {};
+
+    public function addAPIResult(string $type, string $name): void {
```
Use APIDefinitionType enum (new) and switch statements (so that the typechecker makes sure that you handle all the enum values)

fredemmott commented on src/site/SearchResultSet.php:
```diff
@@ -0,0 +1,52 @@
+    private Map<string, string> $functions = Map {};
+    private Map<string, string> $interfaces = Map {};
+    private Map<string, string> $hhvm_guides = Map {};
+    private Map<string, string> $hack_guides = Map {};
```
could be worth doing some typedefs - eg turn this into `Map<DocumentedAPIDefinitionName, SiteURLPath>` ?

fredemmott commented:
much cleaner, thanks :)

jaredonline commented:
Okay, I think I've addressed all the latest round of concerns [=

And I removed the :hankey: talk =p

fredemmott commented:
Going to do some cleanup/changes, will ref this issue
